### PR TITLE
[libc++][memory] P1132R8: `out_ptr` - a scalable output pointer abstraction

### LINKED ISSUE #105001: LWG3594: `inout_ptr` — inconsistent `release()` in destructor
**Link:** https://wg21.link/LWG3594

### LINKED ISSUE #105105: LWG3734: Inconsistency in `inout_ptr` and `out_ptr` for empty case
**Link:** https://wg21.link/LWG3734

### LINKED ISSUE #105146: P1132R8: out_ptr - a scalable output pointer abstraction
**Link:** https://wg21.link/P1132R8

### LINKED ISSUE #105297: LWG3897: `inout_ptr` will not update raw pointer to 0
**Link:** https://wg21.link/LWG3897

### PR DESCRIPTION
**Pull request by H-G-Hristov**
Differential Revision: https://reviews.llvm.org/D150525

Implements:
- https://wg21.link/P1132R8 - `out_ptr` - a scalable output pointer abstraction
  - https://eel.is/c++draft/smartptr.adapt - 20.3.4 Smart pointer adaptors
- https://wg21.link/LWG3734 - Inconsistency in `inout_ptr` and `out_ptr` for empty case
- https://wg21.link/LWG3897- `inout_ptr` will not update raw pointer to 0

### CODE REVIEW

File: libcxx/docs/Status/Cxx23Papers.csv [hunk 1]
```diff
@@ -13,7 +13,7 @@
 "","","","","","",""
 "`P0401R6 <https://wg21.link/P0401R6>`__","LWG","Providing size feedback in the Allocator interface","June 2021","|Complete|","15.0"
 "`P0448R4 <https://wg21.link/P0448R4>`__","LWG","A strstream replacement using span<charT> as buffer","June 2021","",""
-"`P1132R8 <https://wg21.link/P1132R8>`__","LWG","out_ptr - a scalable output pointer abstraction","June 2021","",""
+"`P1132R8 <https://wg21.link/P1132R8>`__","LWG","out_ptr - a scalable output pointer abstraction","June 2021","|Complete|","17.0"
```
Comment (ldionne) [inline libcxx/docs/Status/Cxx23Papers.csv:16]:
```suggestion
"`P1132R8 <https://wg21.link/P1132R8>`__","LWG","out_ptr - a scalable output pointer abstraction","June 2021","|Complete|","18.0"
```

File: libcxx/include/__memory/pointer_traits.h [hunk 2]
```diff
@@ -15,11 +15,14 @@
 #include <__type_traits/conditional.h>
 #include <__type_traits/conjunction.h>
 #include <__type_traits/decay.h>
+#include <__type_traits/enable_if.h>
```
Comment (ldionne) [inline libcxx/include/__memory/pointer_traits.h:18]: Not needed.

File: libcxx/test/std/utilities/smartptr/adapt/out.ptr/out.ptr.t.compile.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+// <memory>
+
+// [out.ptr.t], class template out_ptr_t
+// template<class Smart, class Pointer, class... Args>
+//   class out_ptr_t;                                          // since c++23
+
+#include <memory>
+
+int main(int, char**) {
```
Comment (ldionne) [inline libcxx/test/std/utilities/smartptr/adapt/out.ptr/out.ptr.t.compile.pass.cpp:19]: What is this trying to test? Please add a comment explaining what this is testing.

Also, I don't think there's any reason why we shouldn't be running this test, so it should probably just be in `out.ptr.pass.cpp`.

File: libcxx/include/__memory/out_ptr.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+template <class _Tp>
+concept __resettable_adapted_ptr = requires(_Tp __ptr) { __ptr().reset(); };
+
+template <class _Smart, class _Pointer, class... _Args>
+class _LIBCPP_TEMPLATE_VIS out_ptr_t {
+  static_assert(!__is_specialization_v<_Smart, shared_ptr> || sizeof...(_Args) > 0,
+                "Specialization of std::shared_ptr<> requires a deleter.");
+
+public:
+  _LIBCPP_HIDE_FROM_ABI explicit out_ptr_t(_Smart& __s, _Args... __args)
```
Comment (ldionne) [inline libcxx/include/__memory/out_ptr.h:39]: Do you have a test that this constructor is explicit? I don't see any.

BTW there should be roughly one test file per public method, which means there should be one for this constructor. Isolating methods this way is a bit tedious, but it makes sure that it's absolutely trivial to tell what is being tested and what isn't. In that case I could have headed to `out_ptr.ctor.pass.cpp` and checked that there's a line saying `static_assert(noexcept(...))`.

And then you can have something like `out_ptr.general.pass.cpp` where you test general usage of the class, which seems to be roughly what you have in `out.ptr.pass.cpp` right now.

This comment applies to `inout_ptr` as well.

File: libcxx/include/__memory/out_ptr.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+private:
+  _Smart& __s_;
+  tuple<_Args...> __a_;
+  _Pointer __p_ = _Pointer();
+};
+
+template <class _Pointer = void, class _Smart, class... _Args>
+_LIBCPP_HIDE_FROM_ABI auto out_ptr(_Smart& __s, _Args&&... __args) {
+  using _Ptr = conditional_t<is_void_v<_Pointer>, __pointer_of_t<_Smart>, _Pointer>;
+  return std::out_ptr_t<_Smart, _Ptr, _Args&&...>{__s, std::forward<_Args>(__args)...};
```
Comment (ldionne) [inline libcxx/include/__memory/out_ptr.h:90]: The wording uses `out_ptr_t<...>(...)` instead of `out_ptr_t<...>{...}` -- let's stick with that just for consistency.

File: libcxx/include/__memory/out_ptr.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    } else if constexpr (is_constructible_v<_Smart, _SP, _Args...>) {
+      std::apply([&](auto&&... __args) { __s_ = _Smart(static_cast<_SP>(__p_), std::forward<_Args>(__args)...); },
+                 std::move(__a_));
+    } else {
+      static_assert(__resettable_smart_pointer_with_args<_Smart, _Pointer, _Args...> ||
+                    is_constructible_v<_Smart, _SP, _Args...>);
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI operator _Pointer*() const noexcept { return std::addressof(const_cast<_Pointer&>(__p_)); }
```
Comment (ldionne) [inline libcxx/include/__memory/out_ptr.h:71]: You don't seem to have any tests for these methods exclusively. In particular, you're not testing that these are `noexcept` and `const`. I'm pretty sure all of the tests will pass if you remove `const` here at the moment.

**Review by ldionne — CHANGES_REQUESTED** [review]
Thanks for re-uploading this patch from Phabricator! I didn't do an in-depth review of the implementation yet but I do have some comments. I'll do another pass once they have been addressed.

File: libcxx/include/__memory/out_ptr.h [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+private:
+  _Smart& __s_;
+  tuple<_Args...> __a_;
+  _Pointer __p_ = _Pointer();
+};
+
+template <class _Pointer = void, class _Smart, class... _Args>
+_LIBCPP_HIDE_FROM_ABI auto out_ptr(_Smart& __s, _Args&&... __args) {
+  using _Ptr = conditional_t<is_void_v<_Pointer>, __pointer_of_t<_Smart>, _Pointer>;
```
Comment (ldionne) [inline libcxx/include/__memory/out_ptr.h:89]: I think there's a pedantic bug here. If fetching `__pointer_of_t<_Smart>` is a hard error but we specify a `_Pointer` type explicitly, the current implementation is going to fail to compile because we always eagerly evaluate `__pointer_of_t<_Smart>`. But I think it should compile successfully per the standard.

Edit: I'm unable to find a way to trigger that, since the typical way would involve making a smart ptr that contains a hard-error when you instantiate it to get the nested `::pointer` type. But that strategy obviously doesn't work here. So maybe this isn't a problem after all.

File: libcxx/include/__memory/inout_ptr.h [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#if _LIBCPP_STD_VER >= 23
+
+template <class _Smart, class _Pointer, class... _Args>
+class _LIBCPP_TEMPLATE_VIS inout_ptr_t {
+  static_assert(!__is_specialization_v<_Smart, shared_ptr>, "std::shared_ptr<> is not supported");
+
+public:
+  _LIBCPP_HIDE_FROM_ABI explicit inout_ptr_t(_Smart& __s, _Args... __args)
```
Comment (ldionne) [inline libcxx/include/__memory/inout_ptr.h:36]: The current naming is a bit confusing IMO, I would rename `__s` to something that's less similar to `__s_`. Anything will do, really.

File: libcxx/include/__memory/inout_ptr.h [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+            return __s;
+          } else {
+            return __s.get();
+          }
+        }()) {
+    if constexpr (requires { __s.release(); }) {
+      __s.release();
+    } else {
+      __s = _Smart();
+    }
```
Comment (ldionne) [inline libcxx/include/__memory/inout_ptr.h:48]: Now here you actually want to use `__s_`, not the argument. It's equivalent because they're references but semantically what you're doing is resetting your internal smart pointer, not the argument.

File: libcxx/include/__memory/inout_ptr.h [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  _LIBCPP_HIDE_FROM_ABI inout_ptr_t(const inout_ptr_t&) = delete;
+
+  _LIBCPP_HIDE_FROM_ABI ~inout_ptr_t() {
+    if constexpr (!is_pointer_v<_Smart>) {
```
Comment (ldionne) [inline libcxx/include/__memory/inout_ptr.h:54]: Does this `if constexpr` have a test?

File: libcxx/include/__memory/inout_ptr.h [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+private:
+  _Smart& __s_;
+  tuple<_Args...> __a_;
+  _Pointer __p_;
+};
+
+template <class _Pointer = void, class _Smart, class... _Args>
+_LIBCPP_HIDE_FROM_ABI auto inout_ptr(_Smart& __s, _Args&&... __args) {
+  using _Ptr = conditional_t<is_void_v<_Pointer>, __pointer_of_t<_Smart>, _Pointer>;
+  return std::inout_ptr_t<_Smart, _Ptr, _Args&&...>(__s, std::forward<_Args>(__args)...);
```
Comment (ldionne) [inline libcxx/include/__memory/inout_ptr.h:95]: If you do something like

```
struct Foo {
  Foo() = default;
  Foo(Foo const&) = delete; // not copyable
  Foo(Foo&&) = delete; // not movable either!
};

Foo arg;
std::inout_ptr(smart, arg);
```

It should work. Your current implementation will work, but it would break if you didn't store reference in your tuple, or if you used `std::inout_ptr_t<_Smart, _Ptr, std::decay_t<_Args>...>`. Basically, we can add a test like this to ensure that you're storing references in your tuple internally.

Applies to both `inout_ptr` and `out_ptr`.

File: libcxx/test/std/utilities/smartptr/adapt/inout.ptr/inout.ptr.general.pass.cpp [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+
+void replace_SomeInt_void_p(void** pp) {
+  assert(reinterpret_cast<SomeInt*>(*pp)->value == 90);
+  delete static_cast<SomeInt*>(*pp);
+  *pp = reinterpret_cast<void*>(new SomeInt{9084});
+}
+
+// Test `std::inout_ptr()` function.
+
+void test_raw_ptr() {
```
Comment (ldionne) [inline libcxx/test/std/utilities/smartptr/adapt/inout.ptr/inout.ptr.general.pass.cpp:73]: I would suggest the following organization instead:

```c++
void test() {
  // Test with <explain in plain english>
  {
    auto f(int** pp) -> void {
      assert(**pp == 90);
      delete *pp;
      *pp = new int{84};
    };

    // raw pointer
    {
      int* ptr = new int{90};
      f(std::inout_ptr<int*>(ptr));
      assert(*ptr == 84);
      delete ptr;
    }
    // unique_ptr
    {
      std::unique_ptr<int> ptr = std::make_unique<int>(90);
      f(std::inout_ptr<int*>(ptr));
      assert(*ptr == 84);
    }
  }

  // ... same test for other C style functions ...
}
```

This localizes what you're testing a bit more and IMO it's easier to read. WDYT?

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/include/__memory/out_ptr.h [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+#if _LIBCPP_STD_VER >= 23
+
+template <class _Tp>
+concept __resettable_adapted_ptr = requires(_Tp __ptr) { __ptr().reset(); };
+
+template <class _Smart, class _Pointer, class... _Args>
+class _LIBCPP_TEMPLATE_VIS out_ptr_t {
+  static_assert(!__is_specialization_v<_Smart, shared_ptr> || sizeof...(_Args) > 0,
+                "Specialization of std::shared_ptr<> requires a deleter.");
```
[the remark this answers is not in the document]
Comment (H-G-Hristov) [reply libcxx/include/__memory/out_ptr.h:36]: @ldionne After rereading the specs I found this requirements, did we miss them?

Screenshot 2023-12-14 at 21 20 25
Screenshot 2023-12-14 at 21 28 04

My interpretation is that the standard mandates:

```c++
template <class _Smart, class _Pointer, class... _Args>
class _LIBCPP_TEMPLATE_VIS out_ptr_t {
  static_assert(!__is_specialization_v<_Smart, shared_ptr> || sizeof...(_Args) > 0,
                "Specialization of std::shared_ptr<> requires a deleter.");

template <class _Smart, class _Pointer, class... _Args>
class _LIBCPP_TEMPLATE_VIS inout_ptr_t {
  static_assert(!__is_specialization_v<_Smart, shared_ptr>, "std::shared_ptr<> is not supported");
```
Also implemented similarly in libstdc++ with more clear message than mine: 
https://github.com/gcc-mirror/gcc/blob/95b70545331764c85079a1d0e1e19b605bda1456/libstdc%2B%2B-v3/include/bits/out_ptr.h#L57
https://github.com/gcc-mirror/gcc/blob/95b70545331764c85079a1d0e1e19b605bda1456/libstdc%2B%2B-v3/include/bits/out_ptr.h#L296

File: libcxx/include/__memory/inout_ptr.h (hunk 11 above)
Comment (H-G-Hristov) [reply libcxx/include/__memory/inout_ptr.h:94]: Maybe I don't understand but the following fails to compile: https://godbolt.org/z/cseMEfdj5

> opt/compiler-explorer/gcc-trunk-20240117/include/c++/14.0.1/bits/out_ptr.h:212:42: error: use of deleted function 'constexpr Foo& Foo::operator=(const Foo&)'
  212 |               _M_smart._M_t._M_deleter() = std::forward<_Del2>(_M_del);
      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```c++
#include <memory>
#include <tuple>

struct Foo {
    Foo() = default;
    Foo(Foo const&) = delete;  // not copyable
    Foo(Foo&&) = delete;       // not movable either!

    void operator()(int*) {};
};

int main() {
    std::unique_ptr<int, Foo> smart;
    Foo arg;
    // std::ignore = std::out_ptr(smart, arg);
    std::ignore = std::inout_ptr(smart, arg);
}
```

File: libcxx/docs/Status/Cxx23Issues.csv [hunk 14]
```diff
@@ -282,6 +282,8 @@
 "`3645 <https://wg21.link/LWG3645>`__","``resize_and_overwrite`` is overspecified to call its callback with lvalues","February 2023","|Complete|","14.0",""
 "`3655 <https://wg21.link/LWG3655>`__","The ``INVOKE`` operation and union types","February 2023","|Complete|","18.0",""
 "`3723 <https://wg21.link/LWG3723>`__","``priority_queue::push_range`` needs to ``append_range``","February 2023","","","|ranges|"
+"`3734 <https://wg21.link/LWG3734>`__","Inconsistency in ``inout_ptr`` and ``out_ptr`` for empty case","February 2023","|Complete|","18.0",""
+"`3772 <https://wg21.link/LWG3772>`__","``repeat_view``'s ``piecewise`` constructor is missing Postconditions","February 2023","","","|ranges|"
 "`3734 <https://wg21.link/LWG3734>`__","Inconsistency in ``inout_ptr`` and ``out_ptr`` for empty case","February 2023","","",""
 "`3772 <https://wg21.link/LWG3772>`__","``repeat_view``'s ``piecewise`` constructor is missing Postconditions","February 2023","|Complete|","17.0","|ranges|"
```
Comment (H-G-Hristov) [inline libcxx/docs/Status/Cxx23Issues.csv:288]:
```suggestion
"`3734 <https://wg21.link/LWG3734>`__","Inconsistency in ``inout_ptr`` and ``out_ptr`` for empty case","February 2023","|Complete|","19.0",""
"`3772 <https://wg21.link/LWG3772>`__","``repeat_view``'s ``piecewise`` constructor is missing Postconditions","February 2023","|Complete|","17.0","|ranges|"
```

File: libcxx/include/libcxx.imp [hunk 15]
```diff
@@ -0,0 +1,874 @@
+[
```
Comment (ldionne) [inline libcxx/include/libcxx.imp:1]: This file isn't checked in anymore, it should be removed.

File: libcxx/include/__memory/out_ptr.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/include/__memory/out_ptr.h:38]: Ack, I was wrong.

File: libcxx/include/__memory/out_ptr.h [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+#include <__undef_macros>
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#if _LIBCPP_STD_VER >= 23
+
+template <class _Smart, class _Pointer, class... _Args>
+class _LIBCPP_TEMPLATE_VIS out_ptr_t {
+  static_assert(!__is_specialization_v<_Smart, shared_ptr> || sizeof...(_Args) > 0,
+                "Specialization of std::shared_ptr<> requires a deleter.");
```
Comment (ldionne) [inline libcxx/include/__memory/out_ptr.h:38]:
```suggestion
                "Using std::shared_ptr<> without a deleter in std::out_ptr is not supported.");
```

Or

```suggestion
                "Using std::shared_ptr<> in std::out_ptr requires a deleter.");
```

File: libcxx/include/__memory/out_ptr.h [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+public:
+  _LIBCPP_HIDE_FROM_ABI explicit out_ptr_t(_Smart& __smart, _Args... __args)
+      : __s_(__smart), __a_(std::forward<_Args>(__args)...), __p_() {
+    using _Ptr = decltype(__smart);
+    if constexpr (__resettable_smart_pointer<_Ptr>) {
+      __s_.reset();
+    } else if constexpr (is_constructible_v<_Smart>) {
+      __s_ = _Smart();
+    } else {
+      static_assert(__resettable_smart_pointer<_Ptr> || is_constructible_v<_Smart>);
```
Comment (ldionne) [inline libcxx/include/__memory/out_ptr.h:49]: An error message would be nice.

File: libcxx/include/__memory/inout_ptr.h [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+_LIBCPP_PUSH_MACROS
+#include <__undef_macros>
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#if _LIBCPP_STD_VER >= 23
+
+template <class _Smart, class _Pointer, class... _Args>
+class _LIBCPP_TEMPLATE_VIS inout_ptr_t {
+  static_assert(!__is_specialization_v<_Smart, shared_ptr>, "std::shared_ptr<> is not supported");
```
Comment (ldionne) [inline libcxx/include/__memory/inout_ptr.h:38]:
```suggestion
  static_assert(!__is_specialization_v<_Smart, shared_ptr>, "std::shared_ptr<> is not supported with std::inout_ptr");
```

File: libcxx/test/std/utilities/smartptr/adapt/inout_ptr/inout_ptr.general.pass.cpp [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+// template<class Pointer = void, class Smart, class... Args>
+//   auto inout_ptr(Smart& s, Args&&... args);                 // since c++23
+
+#include <cassert>
+#include <memory>
+#include <utility>
+
+#include "../types.h"
+
+void test_replace_int_p() {
```
Comment (ldionne) [inline libcxx/test/std/utilities/smartptr/adapt/inout_ptr/inout_ptr.general.pass.cpp:23]: Please include a plain-english comment explaining these test cases. It's OK to put them in separate functions, but the function name isn't really enough to convey what's being tested here, a plain english comment is helpful.

File: libcxx/test/std/utilities/smartptr/adapt/out_ptr/out_ptr_t.ctor.pass.cpp [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+#include <memory>
+
+#include "test_convertible.h"
+#include "../types.h"
+
+int main(int, char**) {
+  {
+    std::unique_ptr<int> uPtr;
+
+    std::out_ptr_t<std::unique_ptr<int>, int*>{uPtr};
```
Comment (ldionne) [inline libcxx/test/std/utilities/smartptr/adapt/out_ptr/out_ptr_t.ctor.pass.cpp:28]: It feels like we should have a runtime assertion to test the state of the pointer after construction. Does that make sense?

File: libcxx/test/std/utilities/smartptr/adapt/types.h [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+  MoveOnlyDeleter& operator=(MoveOnlyDeleter&&) = default;
+
+  void operator()(T* p) const { delete p; }
+
+  bool wasMoveInitilized = false;
+};
+
+// Custom pointer types.
+
+template <typename _Tp>
```
Comment (ldionne) [inline libcxx/test/std/utilities/smartptr/adapt/types.h:32]:
```suggestion
template <typename T>
```

Here and below. We shouldn't use reserved names in the test suite.

**Review by ldionne — COMMENTED** [review]
Looks mostly good, still a few comments.

File: libcxx/test/std/utilities/smartptr/adapt/inout_ptr/inout_ptr.general.pass.cpp (hunk 19 above)
Comment (Zingam) [reply libcxx/test/std/utilities/smartptr/adapt/inout_ptr/inout_ptr.general.pass.cpp:25]: Done! Thanks that was a hard one. Please check my "plain English comments" if they make any sense.

File: libcxx/include/__memory/out_ptr.h (hunk 17 above)
Comment (Zingam) [reply libcxx/include/__memory/out_ptr.h]: DONE. Added error messages.

File: libcxx/test/std/utilities/smartptr/adapt/types.h [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+  explicit ConstructiblePtr(T* p) : ptr{p} {}
+
+  auto operator==(T val) { return *ptr == val; }
+
+  auto* get() const { return ptr.get(); }
+
+  void release() { ptr.release(); }
+};
+
+static_assert(std::is_same_v<std::__pointer_of_t< ConstructiblePtr<int>>, int* >);
```
Comment (ldionne) [inline libcxx/test/std/utilities/smartptr/adapt/types.h:47]: Here and below, we can't use libc++ private helpers cause the test suite needs to work with other implementations as well. I would either remove this `static_assert` or turn it into `LIBCPP_STATIC_ASSERT` if it's really really important to check.

File: libcxx/test/std/utilities/smartptr/adapt/out_ptr/out_ptr_t.ctor.pass.cpp (hunk 20 above)
Comment (H-G-Hristov) [reply libcxx/test/std/utilities/smartptr/adapt/out_ptr/out_ptr_t.ctor.pass.cpp:29]: I think we have full indirect tests for these cases in *.general.pass.cpp. I added a minimal runtime assertion as requested and a comment referencing that fact.

**Review by ldionne — APPROVED** [review]
LGTM with my final comments applied and CI passing. Thanks!

**Comment by H-G-Hristov** [comment]
> LGTM with my final comments applied and CI passing. Thanks!

@ldionne Thank you very much!

File: libcxx/test/std/utilities/smartptr/adapt/types.h (hunk 22 above)
Comment (H-G-Hristov) [reply libcxx/test/std/utilities/smartptr/adapt/types.h]: I think I'd like to have that test. I'll convert it to `LIBCPP_STATIC_ASSERT`. Thanks!

File: libcxx/include/__memory/inout_ptr.h [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  _LIBCPP_HIDE_FROM_ABI operator _Pointer*() const noexcept { return std::addressof(const_cast<_Pointer&>(__p_)); }
+
+  _LIBCPP_HIDE_FROM_ABI operator void**() const noexcept
+    requires(!is_same_v<_Pointer, void*>)
+  {
+    static_assert(is_pointer_v<_Pointer>, "The conversion to void** requires _Pointer to be a raw pointer.");
+
+    return reinterpret_cast<void**>(static_cast<_Pointer*>(*this));
```
Comment (davidben) [inline libcxx/include/__memory/inout_ptr.h:88]: Drive-by comment: Is this cast actually strictly legal? See https://github.com/llvm/llvm-project/issues/133445
